### PR TITLE
Add new Lint/ScriptPermission cop

### PR DESCRIPTION
**Pull request by yhirano55**
This cop checks if a file which has a shebang line as
its first line is granted execute permission.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [ ] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: config/enabled.yml [hunk 1]
```diff
@@ -868,6 +868,13 @@ Style/SafeNavigation:
                   safe navigation (`&.`).
   Enabled: true
 
+Style/ScriptPermission:
+  Description: 'Set file permission with `chmod +x`.'
```
Comment (bbatsov) [inline config/enabled.yml]: I'd use  some more generic wording here - after all some people use Windows. ;-)

File: lib/rubocop/cop/style/script_permission.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop checks whether the file permission is executable.
```
Comment (bbatsov) [inline lib/rubocop/cop/style/script_permission.rb]: This should be extended a bit - after all the cop applies just scripts. Frankly, from this description people looking the manual would be quite confused.

**Comment by bbatsov** [comment]
This cop should certain check that the script files actually have shebang in them. Right now this is going to ignore every scripts that's not in `bin/` and `exe/` and that sucks. 

Guess some logic can be used from other cops that we have which are dealing with shebangs.

**Comment by yhirano55** [comment]
@bbatsov Thanks for your code-review! I have addressed your feedbacks. Please review this codes, if you have a chance :)

File: lib/rubocop/cop/style/script_permission.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Style
+      # This cop checks if a file which has a shebang line as
+      # its first line is granted execute permission.
+      class ScriptPermission < Cop
+        MSG = 'Grant script file execute permission.'.freeze
```
Comment (bbatsov) [inline lib/rubocop/cop/style/script_permission.rb]: Maybe we should mention the file name here, as this wording might convince some people. I'd also use working like `Script file X doesn't have execute permission.` (or something similar).

**Comment by bbatsov** [comment]
Please, update the specs to use the new rspec helpers recently added by @backus.

File: lib/rubocop/cop/style/script_permission.rb (hunk 3 above)
Comment (yhirano55) [inline lib/rubocop/cop/style/script_permission.rb]: Oh, I see. I'll fix it. Thanks!!

**Comment by backus** [comment]
This should probably go under the `Lint` namespace. Doesn't seem like style

**Comment by yhirano55** [comment]
@bbatsov @backus Thanks for your comments. I've done:

- Used new rspec helpers.
- Changed department from `Style` to `Lint`.

**Comment by zavan** [comment]
I'm having a problem with this cop, it says my `bin/update` file doesn't have execute permissions, but it does.

I have tried `chmod +x bin/update` to no effect, here's my `ls -laH` output:
```
zavan@zavan-pc:~/projects/ruby/alm$ ls -laH bin/
total 92
drwxr-xr-x  2 zavan zavan 4096 mai 30 14:12 .
drwxr-xr-x 16 zavan zavan 4096 mai 30 13:54 ..
-rwxr-xr-x  1 zavan zavan  129 fev 24 20:41 bundle
-rwxr-xr-x  1 zavan zavan  180 jan 30 00:34 delayed_job
-rwxr-xr-x  1 zavan zavan  141 mai 22 23:46 rails
-rwxr-xr-x  1 zavan zavan   90 mai 22 23:46 rake
-rwxr-xr-x  1 zavan zavan  376 jan 30 00:34 rspec
-rwxr-xr-x  1 zavan zavan  920 fev 24 22:03 setup
-rwxr-xr-x  1 zavan zavan  782 mai 30 14:12 update
```
Notice that all files have the same permissions, but only this one is erroring out.

I have checked and the first line is `#!/usr/bin/env ruby`, here's the full source code:
```ruby
#!/usr/bin/env ruby
require 'pathname'
require 'fileutils'
include FileUtils

# path to your application root.
APP_ROOT = Pathname.new File.expand_path('../../', __FILE__)

def system!(*args)
  system(*args) || abort("\n== Command #{args} failed ==")
end

chdir APP_ROOT do
  # This script is a way to update your development environment automatically.
  # Add necessary update steps to this file.

  puts '== Installing dependencies =='
  system! 'gem install bundler --conservative'
  system('bundle check') || system!('bundle install')

  puts "\n== Updating database =="
  system! 'bin/rails db:migrate'

  puts "\n== Removing old logs and tempfiles =="
  system! 'bin/rails log:clear tmp:clear'

  puts "\n== Restarting application server =="
  system! 'bin/rails restart'
end
```

I have also tried deleting and recreating the file and rewriting the shebang line.

Also, whatever happened to #3659? I agree that this is out of scope.

**Comment by sonalkr132** [comment]
@zavan I think the buffer/cache might be causing the problem. In my case, the cop complained correctly that a script file wasn't executable and it didn't shut up until I changed content of the file (add/delete an empty line)

**Pull request merged**